internal/federationdomain/oidc/oidc.go

```diff
@@ -104,26 +105,152 @@ type UpstreamStateParamData struct {
+	// the ID token lifetime, and the Concierge's mTLS client cert lifetime.
+	// This is because a client can exchange the access token just before it expires
+	// for a new cluster-scoped ID token, and use that just before it expires to get
+	// a new mTLS client cert, which grants access to the cluster until it expires.
+	//
+	// Note that the Concierge's mTLS client cert lifetime is 5 minutes, which can
+	// be seen in its source at credentialrequest/rest.go.
+	//
+	// This maximum total time is important because it represents the longest possible
+	// time that a user could continue to use a cluster based on their original login
+	// (or most recent refresh) after an administrator of an external identity provider
+	// removes the user, revokes their session, changes their group membership,
+	// or otherwise makes any type of change to the user's account in the external
+	// identity provider that should be noticed by the Supervisor during an upstream
+	// refresh.
+	//
+	// Given the timeouts specified below:
+	// For sessions started using a GitHub PAT, this is 8 + 2 + 5 = 15 minutes.
+	// For sessions started any other way, this is 2 + 2 + 5 = 9 minutes.
+	//
+	// The CLI will use a cached mTLS client cert until it expires. For a session
+	// started using a GitHub PAT, when first mTLS client cert has expired, the CLI
+	// will be able to fetch a second one without performing a refresh for 3 more minutes.
+	// If the client is actively making Kubernetes API requests during this time, this
+	// should give at least 10 minutes of cluster access (two mTLS client cert lifetimes).
+	// For any other session type, when the first mTLS client cert expires, the CLI will
+	// need to perform a refresh before it can get a second client cert.
+
+	// Give a generous amount of time for an authorized client to be able to exchange
+	// its authcode for tokens.
 	authorizationCodeLifespan := 10 * time.Minute
+
+	// This is intended to give a very short amount of time to allow the client to
+	// use the access token to exchange for cluster-scoped ID token(s). After this
+	// time runs out, they will need to perform a refresh to get a new tokens,
+	// ensuring the Supervisor has a chance to revalidate their session often.
+	accessTokenLifespan := 2 * time.Minute
+
+	// The ID token can have the same lifespan as the access token. It does not grant
+	// access to anything in a typical Pinniped setup.
+	idTokenLifespan := accessTokenLifespan
+
+	// This is just long enough to cover a typical work day, giving the end user an
+	// experience of logging in once per day to access all their Kubernetes clusters.
 	refreshTokenLifespan := 9 * time.Hour
 
+	// Give a little extra time for some storage lifetimes, to avoid the possibility
+	// that the storage be garbage collected in the middle of trying to look up the token.
+	storageExtraLifetime := time.Minute
+
+	// This is longer than the 5-minute lifetime of mTLS client certs issued by the Concierge,
```

Comment:
Curious if we would want to split out Github specifics into a separate file, or consider if Github specifics belong in default timeout config.  This is likely just a tiny `nit`.

Reply:
Perhaps, although I don't feel like it's too bad for it to live in a central place that decides all the timeouts either. Certainly could be refactored in a real implementation PR if desired.